### ISSUE TITLE
Calling SingleFile from another extension.

### ISSUE THREAD

**Issue by Dalton-Overlin** [gildas-lormeau/SingleFile#515]
Hi, I have a question, can I call SingleFile from a different Firefox extension and have SingleFile save the currently active tab? You see, I am working on an extension that will call SingleFile once a specific HTML element is found in a webpage. But, I cannot find how to initiate SingleFile from a separate extension. Maybe its possible, maybe its not, so I thought I would ask? If it is possible, how would I call SingleFile using Javascript to get it to save the current active page? If you could let me know if SingleFile supports this and if it does, if you could let me know how to call single file I would really appreciate it. Also, if SingleFile cannot be called from a separate extension would it be possible for me to integrate SingleFile into my own extension and if so how would I do that and call SingleFile to save the current active tab? As I've mentioned, I am trying to make an extension that will check for a certain HTML element to be present on a page and then save the page once that HTML element is present. So the extension will not be using browser events such as hotkeys to know when to start the page saving process. But instead, a script in my extension will listen check for a HTML element in a page and once that HTML element is found it should use SingleFile to save the currently active tab.  

Sincerest regards.

**Comment by livelace** [comment]
Well, I need this feature too :) What's the best way to call SingleFile and get content as a string ? :)

**Comment by gildas-lormeau** [comment]
You could call SingleFile from another extension by using [`runtime.onMessageExternal`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessageExternal). SingleFile already uses this API as shown below.
https://github.com/gildas-lormeau/SingleFile/blob/30903d89bb602a22b615a46ecbe7ff73ba73dc2c/extension/core/bg/messages.js#L61-L69
However, today it uses it only to enable/disable the auto-save. It means you would have to fork SingleFile and add the logic to save the page.

Alternatively, you could integrate the API of SingleFile in your extension by following this short [guide](https://github.com/gildas-lormeau/SingleFile/wiki/How-to-integrate-the-API-of-SingleFile-into-an-extension). However, the API will return a string and you'll have to write the code that downloads the page.

**Comment by Dalton-Overlin** [comment]
Thank you for your reply, so, suppose I use this code below to call single file from another extension. Could you help me with how this message should be structured so it will work with SingleFile? 

```
browser.runtime.sendMessage(
    "red@mozilla.org",
    "my message"
  );
```

Also, based on your answer, I plan to fork the project (it seems like the best solution). So it is likely that I will be calling SingleFile from within the same extension using that code above. So, do you also have an example for how I can add the logic to save the page once the runtime message is received rather than just enabling/disabling auto-save? For example, in what script must this logic be added and do you have an example of that logic? This way I could use the above runtime message (after setting it up correctly first) and then once that runtime message is received by SingleFile it should begin saving the page once I add the need logic to the forked SingleFile. If you could help me with those few things I would sincerely appreciate it.

Sincerest regards.

**Comment by gildas-lormeau** [comment]
To save the page instead of auto-saving it when receiving an external message, you should replace in the following code from `messages.js` I posted in my previous response
```js
const currentTab = allTabs[0]; 
if (currentTab) { 
	return singlefile.extension.core.bg.autosave.onMessageExternal(message, currentTab, sender); 
} else { 
	return false; 
} 
```
with:
```js
return singlefile.extension.core.bg.business.saveTabs(allTabs);
```

Then, in your extension you should run:
```js
 browser.runtime.sendMessage("{531906d3-e22f-4a6c-a102-8057b88a1a63}", {})
```
The id is defined in the `manifest.json` file.

**Comment by Dalton-Overlin** [comment]
Thank you, @gildas-lormeau this is what I needed to know. I will get busy and make these mods to the forked SingleFile and add that code to my extension. You can close this issue if you would like, it will take me a few days to get the time to implement this project. If I have any further issues with this I will just open another issue. Thank you again for your time and help, I really appreciate it. 

Sincerest regards; Dalton Overlin

**Comment by gildas-lormeau** [comment]
@Dalton-Overlin You're welcome!

@livelace Do you need more info on your end?

**Comment by livelace** [comment]
@gildas-lormeau Not yet, I'm going to finish my apps first and return to this question later.
In my case I use selenium "execute_script", not other extension. Just need to realize how to do it easier (invoke buttons pushing, message sending etc.).   
Thanks for concerns, btw.

**Comment by Dalton-Overlin** [comment]
Hi, @gildas-lormeau I have been working on this project. But, I have hit a dilemma. So, here below is the script I've come up with so far that runs from my extension. Please note that this script below is a "content script" it's not a background script. Now, the problem is that it appears to send the message to the modified SingleFile, like I discussed earlier, but the modified SingleFile doesn't do anything? So I wanted to ask you, have I structured this browser.runtime.sendMessage correctly and can it be run from a content script like I've designed it to do? Again, I want to mention that I made that modification you outlined to the messages.js script in SingleFile. So, I was just wondering if there's something I'm not seeing that you might be able to spot that I'm doing wrong?

`var cuizXVI = '1.0'; // Transparency.
var spanXVI = document.createElement('span');
spanXVI.innerHTML = '<button id="KTVX1997PLUSONE" style="position:absolute;bottom:0;right:0;color:black;background-color:green;opacity: ' + cuizXVI + ';z-index: 1000;">A</button>';
document.body.appendChild(spanXVI);

var buttonXVI = document.getElementById("KTVX1997PLUSONE");

buttonXVI.addEventListener("click", function () {
	if (this.innerText == "A") {
		this.innerText = "B";
	}
});

var intervalXVI = setInterval(function () {
	var buttonXVITW = document.getElementById("KTVX1997PLUSONE");
	if (buttonXVITW.innerText == "B") {
		buttonXVITW.innerText = "A";
		try {
			var sent = browser.runtime.sendMessage("{531906d3-e22f-4a6c-a102-8057b88a1a63}", {});
			console.log("It was sent."+sent);
		} catch (err) {
			console.log("Error");
		}
		buttonXVITW.innerText = "A";
	}
}, 3000);`

**Comment by Dalton-Overlin** [comment]
@gildas-lormeau Also, I wanted to add on to my previous message, I have been working on integrating my scripts with SingleFile directly. This way everything is wrapped in a single extension. In the extension I attached below the JS file named wick.js is the content script I added while the script named wiredIn.js is the background script. I will attach it below. The problem here, is when I attempt to call SingleFile using the command you provided nothing happens? Take note: in this extension I attached below I have placed the code to call SingleFile in the wiredIn.js file. I just wanted to add an update on my progress. Especially now that I have decided to integrate my own scripts directly into SingleFile rather than using a separate extension. The goal, is to get SingleFile to save the currently active web page when the button is clicked that is inserted into every webpage by the wick.js script. In other words, wick.js communicates with wiredIn.js and wiredIn.js should start the current webpage being saved. 
[singlefile-1.18.23-an+fx.zip](https://github.com/gildas-lormeau/SingleFile/files/5436080/singlefile-1.18.23-an%2Bfx.zip)
